keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,644 @@
+        - --hostname-override=${NODE_NAME}
+        env:
+        - NODE_NAME
+          valueFrom:
+            fieldRef:
+              fieldPath: spec.nodeName
+        volumeMounts:
+        - mountPath: /etc/config
+          name: config
+          readOnly: true
+      volumes:
+      - name: config
+        configMap:
+          name: kube-proxy-config-ghkm248tbc
+```
+
+Finally, when instance-specific configuration is also provided via the
+ComponentConfig approach, using the separate file proposed above.
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+  name: kube-proxy
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: instance-config
+        image: something-with-bash
+        command:
+        - bash -c
+        args:
+        - cat << EOF > /etc/config/kube-proxy-instance.yaml
+          kind: InstanceConfiguration
+          apiVersion: kubeproxy.config.k8s.io/v1alpha1
+          hostnameOverride: ${NODE_NAME}
+          EOF
+        env:
+        - NODE_NAME
+          valueFrom:
+            fieldRef:
+              fieldPath: spec.nodeName
+        volumeMounts:
+        - mountPath: /etc/config
+          name: config
+          readOnly: true
+      containers:
+      - name: kube-proxy
+        image: kube-proxy-amd64
+        command:
+        - kube-proxy
+        - --config=/etc/config/kube-proxy.yaml
+        - --instance-config=/etc/config/kube-proxy-instance.yaml
+        volumeMounts:
+        - mountPath: /etc/config
+          name: config
+          readOnly: true
+      volumes:
+      - name: config
+        configMap:
+          name: kube-proxy-config-ghkm248tbc
+```
+
+Although the last example is more verbose, it provides the full benefits of
+the ComponentConfig approach, resulting in a consistent surface and versioning
+policy for the component's parameters.
+
+Sometimes, kube-proxy is configured as a static Pod instead of a DaemonSet.
+In that case, instance-specific configuration could be configured by the node
+startup scripts and mounted via a host path volume.
+
+### Kubelet Example
+
+The Kubelet is not typically deployed as a Pod. Instead, startup scripts that
+bootstrap the node write the Kubelet's initial configuration to disk at startup.
+In some deployments, the Kubelet's configuration may later be modified by the
+Dynamic Kubelet Config feature.
+
+Today, the startup script may write a Kubelet command line that looks something
+like `kubelet --config=/etc/config/kubelet.yaml --node-ip=${NODE_IP}`. Under
+this proposal, it would instead write both config files, and a command line like
+`kubelet --config=/etc/config/kubelet.yaml --instance-config=/etc/config/kubelet-instance.yaml`.
+
+When the Dynamic Kubelet Config feature is enabled, the Kubelet only replaces
+the values supplied to `--config` when consuming a remote configuration. Thus,
+the instance-specific parameters would not interfere with sharable parameters
+in this approach.
```

Comment:
some random toughs:

if a "global" config already exists in the cluster, a component may chose to ignore subsequent "global" configuration input.
this has implications around authz, as future component instances might not be allowed to patch the "global" config and only be allowed to feed "instance" configuration with read access to the "global".

scenario A:
- a higher level tool writes "global" config in the cluster in a CM
- instances of a component can consume "global" but not be allowed to modify it, unless their API server client has the privilege.

scenario B:
- first instance of a component writes the "global" config in a CM
- RBAC privelige for writing the CM is removed by a high level tool
- future instances of the component can only read the CM

Reply:
Both scenarios seem orthogonal to this proposal, because in either case a local override may make sense depending on the component.

In general I would recommend treating the CM as immutable after creation, because global updates are significantly more dangerous than staged rollouts.

Reply:
so potentially the component can error out if the CM already exists and --config is passed again, or should we trust the user to not do that?

Reply:
Maybe you can provide an example to highlight the risk? The way I'm currently looking at it is it's the same user writing both configs, and writing the pod spec, so it seems unlikely they'd make that mistake.

Reply:
But I'm thinking about it from the standpoint where the CM must exist before kubelet will allow the pod to start. If we're talking about optional CM scenarios where the CM might be created by the pod on first startup, I think that gets more complicated and an example would be super-helpful.

Reply:
For instance, how does the component choose to use an initial config vs. the later config from the ConfigMap? Does it read the ConfigMap directly from the API or is the ConfigMap mounted as optional? Is the idea to mount the ConfigMap over the path the initial config was at when it eventually shows up?

I'm leaning towards this use-case being beyond the purpose of this proposal, which is focused on config consumed via file-based APIs rather than via the component calling the Kube API to create or consume config.

Reply:
for podspec managed deployments / DS this seems fine as those are approved by the admin or persona applying the YAML. but for command line level components / services like some sort of a hypotetical overlayed cluster mesh, subsequent mesh node joins can be trusted with less privileges (e.g. "worker" grade nodes). thus, allowing them to override the global CM should be based on the supplied kubeconfig credentials or simply blocking them on an [immutable CM](https://github.com/kubernetes/enhancements/issues/1412).

Reply:
i was thinking that by global we are assuming a CM stored in the cluster, that is always read by new instances of the component.

Reply:
That's an interesting bootstrapping problem. I'd probably go for an Operator to manage the install instead of a self-bootstrap to clearly divide the responsibility.

Some additional thoughts:
- How does one stage the bootstrapping so that the bootstrapper runs first? Does the bootstrapper mount a different ConfigMap that already exists, whereas the workers mount the future ConfigMap as required so they wait until it exists? Does the bootstrapper also deploy the workers (in this case I'd really prefer an Operator for continuous management)?
- It probably makes sense to run the bootstrapper as a different service account than the workers, and not have it participate as a worker, rather than try to drop RBAC after the CM is created, to keep with the overall declarative approach of K8s (temporary RBAC is imperative).
- Self-bootstrap has the advantage that you only need one binary, but it also adds complexity to the worker component. Having a bootstrapper later mutate into a worker also adds complexity to RBAC management, and _potentially_ risks leftover state local to the bootstrapper that a typical worker wouldn't have access to, depending on what the bootstrapper does.